pydicom/dicomdir.py

```diff
@@ -82,54 +105,682 @@ def __init__(self,
+            while record.next:
+                record = record.next
+                self.root.append(record)
+
+        self._records = records
+
+    @property
+    def patient_records(self) -> List[Dataset]:
+        """Return a list of PATIENT directory records.
+
+        Returns
+        -------
+        list of pydicom.dataset.Dataset
+            The ``PATIENT`` type records in the *Directory Record Sequence*.
+        """
+        return [ii.record for ii in self.root if ii.record_type == "PATIENT"]
+
+
+class DirectoryRecord:
+    """Representation of a Directory Record in a DICOMDIR file.
+
+    .. versionadded:: 2.1
+    """
+    def __init__(self,
+                 ds: Dataset,
+                 record: Dataset,
+                 offset: Optional[int] = None) -> None:
+        """Create a new directory record.
+
+        Parameters
+        ----------
+        ds : pydicom.dicomdir.DicomDir
+            The DICOMDIR dataset.
+        record : pydicom.dataset.Dataset
+            The DICOMDIR record.
+        offset : int, optional
+            The byte offset to the record in the DICOMDIR file.
+
+        Attributes
+        ----------
+        children : list of pydicom.dicomdir.DirectoryRecord
+            The child records for the current record.
+        next : pydicom.dicomdir.DirectoryRecord or None
+            The next record at the current level of the directory, or ``None``
+            if there is no next record.
+        parent : pydicom.dicomdir.DirectoryRecord or None
+            The parent record, or ``None`` for top-level records.
+        previous : pydicom.dicomdir.DirectoryRecord or None
+            The previous record at the current level of the directory, or
+            ``None`` if there is no previous record.
+        record : pydicom.dataset.Dataset
+            The directory record dataset from the DICOMDIR's Media Storage
+            Directory instance.
+        """
+        self._dicomdir = ds
+        self.record = record
+        self._offset = offset
+
+        self.parent = None
+        self.next = None
+        self.previous = None
+        self.children = []
+
+    @property
+    def key(self) -> str:
+        """Return a unique key for the record as :class:`str`."""
+        if self.record_type == "PATIENT":
+            # PS3.3, Annex F.5.1: Each Patient ID is unique within a File-set
+            return self.record.PatientID
+        elif self.record_type == "STUDY" and "StudyInstanceUID" in self.record:
+            # PS3.3, Annex F.5.2: Type 1C
+            return self.record.StudyInstanceUID
+        elif self.record_type == "SERIES":
+            return self.record.SeriesInstanceUID
+        elif self.record_type == "PRIVATE":
+            return self.record.PrivateRecordUID
+
+        # PS3.3, Table F.3-3: Required if record references an instance
+        return getattr(
+            self.record, "ReferencedSOPInstanceUIDInFile", str(uuid.uuid4())
+        )
+
```

Comment:
Ah, true. I'll fix this, but it'll probably get re-worked anyway for writing File-sets since I'm not completely happy with the logic surrounding the concept